linkerd/proxy/http/src/detect.rs

```diff
@@ -0,0 +1,188 @@
+                        maybe_h1 = false;
+                    } else {
+                        maybe_h2 = false;
+                    }
+                } else {
+                    trace!("Checking H2 preface");
+                    if &buf[..H2_PREFACE.len()] == H2_PREFACE {
+                        trace!("Matched HTTP/2 prefix");
+                        return Ok(Some(Version::H2));
+                    }
+
+                    // Not a match. Don't check for an HTTP/2 preface again.
+                    maybe_h2 = false;
+                }
+            }
+
+            if maybe_h1 {
+                // Scan up to the first line ending to determine whether the
+                // request is HTTP/1.1. HTTP expects \r\n, so we just look for
+                // any \n to indicate that we can make a determination.
+                if buf[scan_idx..].contains(&b'\n') {
+                    trace!("Found newline");
+                    // If the first line looks like an HTTP/2 first line,
+                    // then we almost definitely got a fragmented first
+                    // read. Only try HTTP/1 parsing if it doesn't look like
+                    // HTTP/2.
+                    if !maybe_h2 {
+                        trace!("Parsing HTTP/1 message");
+                        // If we get to reading headers (and fail), the
+                        // first line looked like an HTTP/1 request; so
+                        // handle the stream as HTTP/1.
+                        if let Ok(_) | Err(httparse::Error::TooManyHeaders) =
+                            httparse::Request::new(&mut [httparse::EMPTY_HEADER; 0]).parse(&buf[..])
+                        {
+                            trace!("Matched HTTP/1");
+                            return Ok(Some(Version::Http1));
+                        }
+                    }
+
+                    // We found the EOL and it wasn't an HTTP/1.x request;
+                    // stop scanning and don't scan again.
+                    maybe_h1 = false;
+                }
+
+                // Advance our scan index to the end of buffer so the next
+                // iteration starts scanning where we left off.
+                scan_idx = buf.len() - 1;
+            }
+
+            if !maybe_h1 && !maybe_h2 {
+                trace!("Not HTTP");
+                return Ok(None);
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bytes::BufMut;
+    use tokio_test::io;
+
+    const HTTP11_LINE: &'static [u8] = b"GET / HTTP/1.1\r\n";
+    const H2_AND_GARBAGE: &'static [u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\ngarbage";
+    const GARBAGE: &'static [u8] =
+        b"garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage garbage";
+
+    #[tokio::test(flavor = "current_thread")]
+    async fn h2() {
+        let _ = tracing_subscriber::fmt().with_test_writer();
```

Comment:
this doesn't do anything. you're constructing a subscriber but not setting it as the default, so it doesn't get used. you want:
```suggestion
        let _guard = tracing_subscriber::fmt().with_test_writer().set_default();
```
(which requires the `SubscriberInitExt` trait in scope, imported by `use tracing_subscriber::prelude::*`)
or 
```suggestion
        let _guard = tracing::subscriber::set_default(
            tracing_subscriber::fmt().with_test_writer().finish()
        );
```
(formatting mine; it's possible that rustfmt won't like that)
or 
```suggestion
    let _ = tracing_subscriber::fmt().with_test_writer().try_init();
```
to try to set it as the global default, which might be simplest if isolating these tests doesn't matter.

(sorry for options overload x_x)